crates/core/src/diff/diff.rs

```diff
@@ -1,523 +1,703 @@
+    }
 }
-impl Eq for Cursor {}
-impl PartialEq for Cursor {
-    #[inline]
-    fn eq(&self, other: &Self) -> bool {
-        self.path.eq(&other.path)
+
+impl<'a> Iterator for Cursor<'a> {
+    type Item = ();
+
+    fn next(&mut self) -> Option<()> {
+        self.advance(false)
     }
 }
-impl PartialOrd for Cursor {
-    #[inline]
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        Some(self.cmp(other))
+
+impl<'a> From<&'a Document> for Cursor<'a> {
+    fn from(doc: &'a Document) -> Self {
+        Cursor::new(doc, doc.root())
     }
 }
-impl Ord for Cursor {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        use std::cmp::Ordering;
 
-        let mut anext = self.path.iter();
-        let mut bnext = other.path.iter();
+impl<'a> Deref for Cursor<'a> {
+    type Target = Node;
 
-        loop {
-            match (anext.next(), bnext.next()) {
-                (None, None) => break,
-                (None, _) => return Ordering::Less,
-                (_, None) => return Ordering::Greater,
-                (a, b) => match a.cmp(&b) {
-                    Ordering::Equal => continue,
-                    other => return other,
-                },
+    fn deref(&self) -> &Self::Target {
+        self.node()
+    }
+}
+
+trait CompatibleWith: Deref<Target = Node> {
+    fn is_compatible_with<T>(&self, other: &T) -> bool
+    where
+        T: Deref<Target = Node>,
+    {
+        match (self.deref(), other.deref()) {
+            (Node::Element(from), Node::Element(to)) => {
+                to.name.eq(&from.name) && to.id().eq(&from.id())
             }
+            (Node::Leaf(_), Node::Leaf(_)) => true,
+            (Node::Root, Node::Root) => true,
+            _ => false,
         }
+    }
+}
+
+impl<T> CompatibleWith for T where T: Deref<Target = Node> {}
 
-        Ordering::Equal
+#[derive(Debug)]
+enum Op<'a> {
+    Continue,
+    /// Detach node and keyed descendants then remove node
+    RemoveNode {
+        node: NodeRef,
+        /// A forked cursor of the node to be removed that iterates over descendant nodes
+        cursor: Cursor<'a>,
+        to: Cursor<'a>,
+        detach: bool,
+    },
+    /// Remove all `from` nodes
+    RemoveNodes {
+        from: Cursor<'a>,
+        to: Cursor<'a>,
+    },
+    /// Append `cursor` relative to the parent node last pushed to the stack by a patch operation
+    Append {
+        from: Cursor<'a>,
+        cursor: Cursor<'a>,
+    },
+    /// Append `to`
+    AppendNodes {
+        from: Cursor<'a>,
+        to: Cursor<'a>,
+    },
+    /// Append sibling nodes relative to the parent node on the stack
+    AppendSiblings {
+        from: Cursor<'a>,
+        cursor: Cursor<'a>,
+    },
+    /// Inserts `node` before the current node and appends descendant nodes
+    InsertBefore {
+        from: Cursor<'a>,
+        cursor: Cursor<'a>,
+    },
+    /// Detach node if not already detached
+    MaybeDetach {
+        node: NodeRef,
+    },
```

Comment:
If a keyed node would be removed and needs to be attached elsewhere it's detached instead, so when it comes time to re-attach, it might have already been detached and if not needs to be.